File: app/src/main/java/com/github/weatherapp/ui/weather/CurrentWeatherActivity.java
package com.github.weatherapp.ui.weather;

import android.app.AlertDialog;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.inputmethod.EditorInfo;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.github.weatherapp.R;
import com.github.weatherapp.injection.component.AppComponent;
import com.github.weatherapp.injection.scope.ActivityScope;
import com.github.weatherapp.ui.base.BaseActivity;
import com.github.weatherapp.ui.settings.SettingsActivity;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.OnEditorAction;
import butterknife.Unbinder;

public class CurrentWeatherActivity extends BaseActivity<CurrentWeatherMvpView, CurrentWeatherMvpPresenter> implements CurrentWeatherMvpView {
    private static final String TAG = CurrentWeatherActivity.class.getSimpleName();

    @BindView(R.id.edit_text_city)
    EditText cityEditText;

    @BindView(R.id.text_city_name)
    TextView cityNameTextView;

    @BindView(R.id.text_temperature)
    TextView temperatureTextView;

    private Component component;

    private Unbinder unbinder;

    @Override
    public void updateCurrentForecast(CurrentWeatherViewModel model) {
        Log.d(TAG, model.toString());
        cityNameTextView.setText(model.getCityName());
        temperatureTextView.setText(model.getTemperature());
        Glide.with(this)
                .load(model.getWeatherIconPath())
                .into((ImageView) findViewById(R.id.image_weather_icon));
    }

    @Override
    public void updateErrorMessage(String message) {
        new AlertDialog.Builder(this)
                .setTitle("Oops!")
                .setMessage("Something went wrong...")
                .show();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        unbinder = ButterKnife.bind(this);
    }

    @Override
    protected void onDestroy() {
        unbinder.unbind();
        super.onDestroy();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.weather, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_settings:
                startActivity(new Intent(this, SettingsActivity.class));
                break;
        }

        return true;
    }

    @Override
    protected CurrentWeatherMvpPresenter newPresenter() {
        return getComponent().presenter();
    }

    @Override
    protected CurrentWeatherMvpView getPresenterView() {
        return this;
    }

    @OnEditorAction(R.id.edit_text_city)
    boolean onEditorActionCityEditText(int actionId) {
        if (actionId == EditorInfo.IME_ACTION_SEARCH) {
            presenter.showCurrentForecast(cityEditText.getText().toString());
            return true;
        }
        return false;
    }

    @OnClick(R.id.button_search)
    void onClickSearchButton() {
        presenter.showCurrentForecast(cityEditText.getText().toString());
    }

    private Component getComponent() {
        if (component == null) {
            component = DaggerCurrentWeatherActivity_Component.builder()
                    .appComponent(getAppComponent())
                    .build();
        }
        return component;
    }

    @ActivityScope
    @dagger.Component(dependencies = AppComponent.class)
    public interface Component {
        CurrentWeatherMvpPresenter presenter();
    }
}

File: app-kotlin/src/main/kotlin/com/github/weatherapp/ui/weather/CurrentWeatherActivity.kt
package com.github.weatherapp.ui.weather

import android.app.AlertDialog
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.inputmethod.EditorInfo
import android.widget.EditText
import android.widget.ImageView
import android.widget.TextView
import butterknife.*
import com.bumptech.glide.Glide
import com.github.weatherapp.R
import com.github.weatherapp.injection.component.AppComponent
import com.github.weatherapp.injection.scope.ActivityScope
import com.github.weatherapp.ui.base.BaseActivity
import com.github.weatherapp.ui.settings.SettingsActivity

class CurrentWeatherActivity : BaseActivity<CurrentWeatherMvpView, CurrentWeatherMvpPresenter>(), CurrentWeatherMvpView {

    @BindView(R.id.edit_text_city)
    internal lateinit var cityEditText: EditText

    @BindView(R.id.text_city_name)
    internal lateinit var cityNameTextView: TextView

    @BindView(R.id.text_temperature)
    internal lateinit var temperatureTextView: TextView

    private val component: Component by lazy {
        DaggerCurrentWeatherActivity_Component.builder()
                .appComponent(appComponent)
                .build()
    }

    private lateinit var unbinder: Unbinder

    override fun updateCurrentForecast(model: CurrentWeatherViewModel) {
        Log.d(TAG, model.toString())
        cityNameTextView.text = model.cityName
        temperatureTextView.text = model.temperature
        Glide.with(this)
                .load(model.weatherIconPath)
                .into(findViewById(R.id.image_weather_icon) as ImageView)
    }

    override fun updateErrorMessage(message: String) {
        AlertDialog.Builder(this)
                .setTitle("Oops!")
                .setMessage("Something went wrong...")
                .show()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        unbinder = ButterKnife.bind(this)
    }

    override fun onDestroy() {
        unbinder.unbind()
        super.onDestroy()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.weather, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_settings -> startActivity(Intent(this, SettingsActivity::class.java))
        }

        return true
    }

    override fun newPresenter(): CurrentWeatherMvpPresenter {
        return component.presenter()
    }

    override val presenterView: CurrentWeatherMvpView
        get() = this

    @OnEditorAction(R.id.edit_text_city)
    fun onEditorActionCityEditText(actionId: Int): Boolean {
        if (actionId == EditorInfo.IME_ACTION_SEARCH) {
            presenter.showCurrentForecast(cityEditText.text.toString())
            return true
        }
        return false
    }

    @OnClick(R.id.button_search)
    fun onClickSearchButton() {
        presenter.showCurrentForecast(cityEditText.text.toString())
    }

    @ActivityScope
    @dagger.Component(dependencies = arrayOf(AppComponent::class))
    interface Component {
        fun presenter(): CurrentWeatherMvpPresenter
    }

    companion object {
        private val TAG = CurrentWeatherActivity::class.java.simpleName
    }
}
